src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java

```diff
@@ -0,0 +1,470 @@
+ */
+public class TileEntityWareHouse extends TileEntityColonyBuilding
+{
+    /**
+     * List which contains the currentTasks to be executed by the deliveryman.
+     */
+    private final CopyOnWriteArrayList<AbstractBuilding> list = new CopyOnWriteArrayList<>();
+
+    /**
+     * Index of last controlled building.
+     */
+    private int index = 1;
+
+    /**
+     * Empty standard constructor.
+     */
+    public TileEntityWareHouse()
+    {
+        super();
+    }
+
+    @NotNull
+    @Override
+    public NBTTagCompound getUpdateTag()
+    {
+        return writeToNBT(new NBTTagCompound());
+    }
+
+    @Override
+    public void update()
+    {
+        super.update();
+
+        if(getColony() != null)
+        {
+            final Map<BlockPos, AbstractBuilding> buildingMap = getColony().getBuildings();
+
+            if (buildingMap.size() < this.index)
+            {
+                this.index = 1;
+            }
+
+            int i = 1;
+            for (@NotNull final Map.Entry<BlockPos, AbstractBuilding> buildingEntry : buildingMap.entrySet())
+            {
+                if (i == index)
+                {
+                    if(buildingEntry.getValue() instanceof AbstractBuildingWorker
+                            && !list.contains(buildingEntry.getValue())
+                            && ((AbstractBuildingWorker) buildingEntry.getValue()).needsAnything())
+                    {
+                        checkInWareHouse((AbstractBuildingWorker) buildingEntry.getValue(), true);
+                    }
+                    this.index++;
+                }
+                i++;
+            }
+        }
+    }
+
+    /**
+     * Get the first task in the list.
+     * @return the building which needs a delivery.
+     */
+    @Nullable
+    public AbstractBuilding getTask()
+    {
+        if(list.isEmpty())
+        {
+            return null;
+        }
+        Log.getLogger().info("Get the task to fullfil it: " + list.get(0).getSchematicName());
+        return list.remove(0);
+    }
+
+    /**
+     * Check if the required items by the building are in the wareHouse.
+     * @param buildingEntry the building requesting.
+     * @param addToList if is in warehouse should add to the list?
+     * @return true if has something in warehouse to deliver.
+     */
+    public boolean checkInWareHouse(@NotNull final AbstractBuilding buildingEntry, boolean addToList)
+    {
+        if(buildingEntry.areItemsNeeded())
+        {
+            for(final ItemStack stack : buildingEntry.getNeededItems())
+            {
+                if(stack == null || (deliveryManHasBuildingAsTask(buildingEntry) && addToList))
+                {
+                    continue;
+                }
+
+                if(isInHut(stack))
+                {
+                    if(addToList)
+                    {
+                        Log.getLogger().info("Add " + buildingEntry.getSchematicName() + " to delivery list");
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal " to delivery list" 4 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1192)